readme.md
[![Build Status](https://jenkins.adc4gis.com/buildStatus/icon?job=ogre)](https://jenkins.adc4gis.com/job/ogre/) [![NPM](https://img.shields.io/npm/v/ogre.svg)](https://npmjs.com/package/ogre) ![NPM Downloads](https://img.shields.io/npm/dt/ogre.svg)

The instructions below are only if you are interested in running the project locally. For help on how to use the web service, visit the [Ogre homepage](http://ogre.adc4gis.com). Ogre makes use of the [ogr2ogr](https://github.com/wavded/ogr2ogr) module.

## Requirements

ogr2ogr requires the command line tool _ogr2ogr_ - [gdal install page](http://trac.osgeo.org/gdal/wiki/DownloadingGdalBinaries). It is recommended to use the latest version.

## Installation

```
npm install -g ogre
```

For specific OS installation help, see the [wiki](https://github.com/wavded/ogre/wiki).

## Running

To run the app:

```
ogre -p 3000
```

And visit the following url in a browser:

```
http://localhost:3000
```

Options include:

```
Usage: ogre [options]

Options:
 -h, --help      help
 -p, --port      port number (default 3000)
 -v, --version   version number
 -t, --timeout   timeout before ogre kills a job in ms (default 15000)
```

## License

(The MIT License)

Copyright (c) 2016 Marc Harter <wavded@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


changelog.md
# 2.0.0 / 2020-02-17

- Breaking change: Support Node >= 10

# 1.3.1 / 2018-10-19

- Upgrade dependencies. Ignore yarn-error
- Limit total upload file size to 100mb

# 1.3.0 / 2016-07-25

- Upgrade dependencies
- Add option to create RFC7946-compliant file for Mapbox upload [dericke]

# 1.2.0 / 2016-07-25

- Add force download option
- Add 400 and 500 status codes [itjope]
- Add optional format option in convertJson [itjope]
- Expose Content-Disposition header [btfou]
- Add extra param to name shp files in zip [btfou]

# 1.1.0 / 2015-06-17

- add; ability to configure timeout

# 1.0.1 / 2015-01-05

- mod; urlencoded limit bump

# 1.0.0 / 2014-11-21

- add; skip failures as an option (not default) [breaking]

# 0.8.0 / 2014-11-04

- deps; update
- mod; switch to jenkins

# 0.7.1 / 2014-06-19

- mod; switch to utf8 responses
- add; docker support
- mod; readme to link to wiki

# 0.6.6 / 2014-01-25

- add; respond to OPTIONS requests with CORS headers to allow preflighted requests [bilts](https://github.com/bilts)

# 0.6.5 / 2014-01-12

- bump; deps (ogr2ogr)

# 0.6.4 / 2014-01-06

- update; deps

# 0.6.3 / 2013-11-20

- add; forcePlainText option

# 0.6.2 / 2013-11-15

- update; ogr2ogr dep

# 0.6.1 / 2013-11-12

- update; ogr2ogr dep

# 0.6.0 / 2013-11-11

- update; broke ogr2ogr wrapper into a separate project
- add; source/target srs reprojection

# 0.5.1 / 2013-09-06

- enable CORS on convert routes [rclark]
- added; npm badge

# 0.5.0 / 2013-07-10

- updated; jade compatability
- updated; adjust temp dir location to use os.tmpDir()
- removed; support for node 0.6
- added; allow a jsonUrl to be POSTed. Get the URL content and then continue

# 0.4.0 / 2013-04-09

- updated; express 3
- updated; app dependencies
- added; node 0.10 travis support

# 0.3.5 / 2013-03-22

- removed; stylus as a dependency

# 0.3.4 / 2012-07-09

- removed; support for 0.4.x
- added; testing for 0.8.x

# 0.3.3 / 2012-04-27

- updated; force express 2.x
- fixed; stdout changed, csv parsing of "

# 0.3.1 / 2012-03-02

- updated; dependencies
- updated; bootstrap

# 0.3.0 / 2012-01-13

- updated; version lockdown for deps
- added; Travis CI

# 0.2.5 / 2011-11-12

- node 0.6 support
- failing test cleanup

# 0.2.4 / 2011-09-17

- Invalid package in package.json

# 0.2.3 / 2011-09-17

- Updated UI look and feel

# 0.2.2 / 2011-09-08

- Added optional output name to geojson to shp
- Added note about convert to shapefile

# 0.2.1 / 2011-08-24

- Strict automatic csv parsing

# 0.2.0 / 2011-08-12

- Added more conversion formats to GeoJSON
- Added convert GeoJSON to Shapefile
- Revamped look and feel


src/ogre.js
var express = require('express'),
    http = require('http'), https = require('https'), url = require('url'),
    toJson = require('./convert-to-json'),
    fromJson = require('./convert-from-json');

var app = null;

function enableCors(req, res, next) {
    res.header("Access-Control-Allow-Origin", "*");
    res.header('Access-Control-Allow-Methods', 'POST');
    res.header("Access-Control-Allow-Headers", "X-Requested-With");

    if (req.method === "OPTIONS") {
        res.send(200);
    } else {
        next();
    }
}

exports.createServer = function(port,maxBuffer,gaCode){
    app = express();
    port || (port = 3000);

    if(maxBuffer) ogre_engine.setMaxBuffer(maxBuffer);

    app.set('views', __dirname + '/views');
    app.set('view engine', 'jade');
    app.use(express.static(__dirname + '/public'))
    app.use(express.bodyParser())

    app.get('/', function(req, res){
        res.render('home', { trackcode: gaCode || '' })
    })

    app.post('/convert', enableCors, function(req, res){
        toJson.upload(req,
            function(outputstream,contentType,launchViewer){
                if(launchViewer)
                    res.render('viewer', { output: outputstream, trackcode: gaCode || '' });
                else {
                    res.header("Content-Type",contentType);
                    res.send(outputstream);
                }
            }
        )
    })

    app.post('/convertJson', enableCors,
     function(req, res, next) {
         if (req.body.jsonUrl) {
            var parsedUrl = url.parse(req.body.jsonUrl),
                prot = parsedUrl.protocol === "http:" ? http : https;
            prot.get(req.body.jsonUrl, function (response) {
                var data = '';
                response.on('data', function (chunk) { data += chunk; });
                response.on('end', function () {
                    req.body.json = data;
                    next();
                });
            });
         } else { next(); }
     },
     function(req, res){
        outputName = (req.body.name || 'ogreToShape') + '.zip';
        fromJson.upload(req,
            function(err, outputZipFile){
                if(err){
                    res.send(err);
                } else {
                    res.download(outputZipFile, outputName, function(err){
                        fromJson.removeOutputFile(outputZipFile);
                    });
                }
            }
        )
    })

    app.use(function(err, req, res, next){
        console.log(err.message, err.stack);
        res.send(err.message, 500);
    })

    app.listen(port);
    console.log("Ogre started, listening on port " + port);

    return true;
}
